Create C style guide

wolf99 opened the pull request:
Initial attempt at the style guide.

Am still unsure on if the capitalization for constants applies only to globals or not.

I intend to add further commits with a short description of the formatting as set by the parameters we pass to `indent` (line length, brace placement, etc)

ryanplusplus commented on docs/C_STYLE_GUIDE.md:
```diff
@@ -0,0 +1,57 @@
+and test files.
+
+The rules within this guide are in addition to, or to call out, those in the 
+[Linux kernal coding style guide](https://www.kernel.org/doc/html/latest/process/coding-style.html).
```
`kernal` => `kernel`

ryanplusplus commented on docs/C_STYLE_GUIDE.md:
```diff
@@ -0,0 +1,57 @@
+
+## Enums
+
+Enum names should usually be singluare rather than plural.
```
`singluare` => `singular`

ryanplusplus commented on docs/C_STYLE_GUIDE.md:
```diff
@@ -0,0 +1,57 @@
+
+Enum names should usually be singluare rather than plural.
+
+Enum labels should be capitalized.
```
Should we call this out as upper snake case?

ryanplusplus commented on docs/C_STYLE_GUIDE.md:
````diff
@@ -0,0 +1,57 @@
+```
+
+Capitalized macro names are appreciated, but macros resembling functions may be
+named in lower case.
````
lower snake case?

ryanplusplus commented on docs/C_STYLE_GUIDE.md:
```diff
@@ -0,0 +1,57 @@
+
+## Constant names
+
+Names for variables defined as `const` should be capitalized.
```
Upper snake case?

ryanplusplus commented:
If the kernel style guide only wants UPPER_SNAKE_CASE for globals, I think we should deviate from them on that point. It feels arbitrary and fiddly.

ryanplusplus commented:
Looks good so far. My only real suggestion is that we should be consistent in how we describe the style (upper snake case vs. just capital/uppercase).

wolf99 commented:
@ryanplusplus thanks for those comments, I have updated with regard to those in f1c282e

I have also added some commentary on the layout of the test file - not sure if this exactly belongs in a style guide but it seems like handy information to have available

ryanplusplus commented on docs/C_STYLE_GUIDE.md:
```diff
@@ -0,0 +1,112 @@
+
+## Enums
+
+Enum names should usually be singluar rather than plural.
```
`singluar` => `singular`

wolf99 replied:
Dammit! I type too fast and then read only what I expect to be there when I check it. Thanks for catching that.

ryanplusplus commented on docs/C_STYLE_GUIDE.md:
```diff
@@ -0,0 +1,112 @@
+# C Style Guide
+
+This guide describes the style intended for use with the C track exercise 
+example and test files.
```
A minor issue: Exercism-wide it was decided to not add manual breaks in lines in Markdown documents. In order to keep things fairly readable (and writable), each sentence should be on its own line. This makes it so that lines don't get forever long in editors that don't soft wrap while also making it so that minor updates to a Markdown document don't require lines to be manually re-wrapped.

This is a fairly recent thing and I'm personally okay with doing no manual wrapping, but if we are going to wrap, let's do it in the recommended way.

wolf99 replied:
Ah, good to know, I'll adjust to suit

ryanplusplus commented:
@wolf99 I think the layout of the test file is definitely helpful. My vote is to keep it.

ryanplusplus commented:
Looks good so far. Are you still planning to add info about `indent`?

wolf99 commented:
Yep, that should be next up

wolf99 commented:
@ryanplusplus  I relised that if non-global constants are to be all-caps,  then that means capitalized names showing up in some unusual places... such as function parameters, e.g.

```c
void some_function_foo(const char *BUFFER, const int BUFFER_LENGTH);
```

To me that looks a bit odd TBH.

ryanplusplus commented:
@wolf99 -- @Gamecock checked the Linux kernel source and didn't find any non-`#define`d, non-enumerated upper snake case constants. So I think we're good to just leave them as they are and go with lower snake case.

wolf99 commented:
@ryanplusplus Have removed the cont style info then. Now this should be good to go, unless you can think of anything else that could be added or adjusted?

Gamecock commented:
New question, but maybe should merge this and save for update two.  Which is correct:

`void some_function_foo(const char *buffer, const int buffer_length);`
or
`void some_function_foo(const char *, const int );`

wolf99 commented:
@Gamecock Good point. At a quick glance the kernel seems to use the style where names are specified. I'd be inclined to stick with that for clarity's sake. Thoughts?

Gamecock commented:
I think that's clearer and we should explicitly require the names be the
same in the .c and .h.

On Oct 24, 2017 12:41 PM, "toby" <notifications@github.com> wrote:

> @Gamecock <https://github.com/gamecock> Good point. At a quick glance the
> kernel seems to use the style where names are specified. I'd be inclined to
> stick with that for clarity's sake. Thoughts?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/c/pull/217#issuecomment-339054143>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFkhSnkqC_FgK8RIagThmYvKY34qqU7Oks5svhM-gaJpZM4P8BF9>
> .
>

ryanplusplus commented:
👍
